lib/auth/auth.ts

```diff
@@ -0,0 +1,137 @@
+import * as ClientOAuth2 from "client-oauth2";
+
+import { TokenManager } from "./tokenManager";
+import { PushNotificationMFA } from "./push";
+import { DeviceBinding } from "./device";
+import { HttpRequest } from "../request";
+
+import { ClientOpts, GetAuthUriOpts } from "../types";
+import { KontistSDKError } from "../errors";
+
+export class Auth {
+  public tokenManager: TokenManager;
+  public device: DeviceBinding;
+  public push: PushNotificationMFA;
+
+  /**
+   * @param baseUrl  Kontist API base url
+   * @param opts     OAuth2 client data including at least clientId, redirectUri,
+   *                 scopes, state and clientSecret or code verifier (for PKCE).
+   * @throws         when both clientSecret and code verifier are provided
+   * @throws         when verifier is provided but state or redirectUri are missing
+   */
+  constructor(baseUrl: string, opts: ClientOpts) {
+    const {
+      clientId,
+      clientSecret,
+      oauthClient,
+      redirectUri,
+      scopes,
+      state,
+      verifier
+    } = opts;
+
+    if (verifier && clientSecret) {
+      throw new KontistSDKError({
+        message:
+          "You can provide only one parameter from ['verifier', 'clientSecret']."
+      });
+    }
+
+    if (verifier && (!state || !redirectUri)) {
+      throw new KontistSDKError({
+        message:
+          "If you are providing a 'verifier', you must also provide 'state' and 'redirectUri' options."
+      });
+    }
+
+    const oauth2Client =
+      oauthClient ||
+      new ClientOAuth2({
+        accessTokenUri: `${baseUrl}/api/oauth/token`,
+        authorizationUri: `${baseUrl}/api/oauth/authorize`,
+        clientId,
+        clientSecret,
+        redirectUri,
+        scopes,
+        state
+      });
+
+    this.tokenManager = new TokenManager(baseUrl, {
+      state,
+      verifier,
+      oauth2Client
+    });
+
+    const request = new HttpRequest(baseUrl, this.tokenManager);
+
+    this.device = new DeviceBinding(this.tokenManager, request);
+    this.push = new PushNotificationMFA(this.tokenManager, request);
+  }
+
+  /**
+   * @deprecated use tokenManager method directly instead
+   *
+   * Build a uri to which the user must be redirected for login.
+   */
+  public getAuthUri = async (opts: GetAuthUriOpts = {}): Promise<string> =>
```

Comment:
Please let me know what you think is the best way to proceed regarding these breaking changes (taking into consideration ease of use for SDK users and maintainability for us), should we:
* add aliases to the `Auth` class like in the current implementation, indicating that they are `@deprecated` (this would prevent anything from breaking, but would discourage usage of the new methods)
* add aliases to the `Auth` class like in the current implementation, **without** indicating that they are deprecated, making the `TokenManager` class more of an implementation detail (downside being that we have to maintain a mapping between `TokenManager` methods and `Auth` aliases)
* do not add aliases here and introduce the breaking changes directly, forcing users to change their code to `client.auth.tokenManager.xxxx` (upside being that there aren't too many users of the SDK currently)

Reply:
As per a discussion with @johannespfeiffer, for now I will keep the `@deprecated` approach adding some warnings in dev mode and specifying when the methods will be removed (thanks for the suggestions 🙏) :
<img width="933" alt="Screenshot 2019-11-12 at 10 34 45" src="https://user-images.githubusercontent.com/3253522/68660011-adc68f00-0538-11ea-875c-186641a7d223.png">


Still open for discussion / other ideas of course.